Point_set_processing_3/include/CGAL/constraint_based_smooth_point_set.h

```diff
@@ -0,0 +1,652 @@
+template <typename Kernel, typename PointRange,
+        typename PointMap, typename VectorMap>
+std::pair<Eigen::Vector3d, Eigen::Matrix3d> calculate_covm_eigenvalues(
+  const typename PointRange::iterator::value_type& vt,
+  PointMap point_map,
+  VectorMap normal_map,
+  const std::vector<typename PointRange::iterator>& neighbor_pwns,
+  typename Kernel::FT normal_threshold,
+  typename Kernel::FT eigenvalue_threshold)
+{
+  // basic geometric types
+  typedef typename Kernel::FT FT;
+  typedef typename Kernel::Vector_3 Vector;
+  typedef typename Kernel::Point_3 Point;
+
+  FT weight = 0;
+  Eigen::Matrix3d covm = Eigen::Matrix3d::Zero();
+  Eigen::Vector3d v_bar = Eigen::Vector3d::Zero();
+
+  const Point& p = get(point_map, vt);
+  const Vector& n = get(normal_map, vt);
+
+  for (typename PointRange::iterator it : neighbor_pwns)
+  {
+    const Point& np = get(point_map, *it);
+    const Vector& nn = get(normal_map, *it);
+
+    Eigen::Vector3d vnp(np.x(), np.y(), np.z());
+
+    FT angle_diff = CGAL::approximate_angle(n, nn);
+    if (angle_diff <= normal_threshold)
+    {
+      weight += 1;
+
+      v_bar += vnp;
+    }
+  }
+
+  if (weight != 0)
+  {
+    v_bar /= weight;
+  }
+
+  for (typename PointRange::iterator it : neighbor_pwns)
+  {
+    const Point& np = get(point_map, *it);
+    const Vector& nn = get(normal_map, *it);
+
+    Eigen::Vector3d vnp(np.x(), np.y(), np.z());
+
+    FT angle_diff = CGAL::approximate_angle(n, nn);
+    if (angle_diff <= normal_threshold)
+    {
+      auto temp_vec = vnp  - v_bar;
+      covm += temp_vec * temp_vec.transpose();
+    }
+  }
+
+  if (weight != 0)
+  {
+    covm /= weight;
+  }
+
+  Eigen::SelfAdjointEigenSolver<Eigen::Matrix3d> solver(covm);
+  std::pair<Eigen::Vector3d, Eigen::Matrix3d> eigens(solver.eigenvalues(), solver.eigenvectors());
+
+  return eigens;
+}
+
+// Calculates new normal positions based off the normal voting tensor
+template <typename Kernel, typename PointRange,
+          typename PointMap, typename VectorMap>
+typename Kernel::Vector_3 nvt_normal_denoising(
+  const typename PointRange::iterator::value_type& vt,
+  PointMap point_map,
+  VectorMap normal_map,
+  const std::pair<Eigen::Vector3d, Eigen::Matrix3d> nvt_eigens,
+  typename Kernel::FT damping_factor)
+{
+  // basic geometric types
+  typedef typename Kernel::FT FT;
+  typedef typename Kernel::Vector_3 Vector;
+  typedef typename Kernel::Point_3 Point;
+
+  Eigen::Matrix3d optimized_nvt = Eigen::Matrix3d::Zero();
+
+  for (int i=0; i<3; ++i)
+  {
+    optimized_nvt += nvt_eigens.first[i] * nvt_eigens.second.col(i) * nvt_eigens.second.col(i).transpose();
+  }
+
+  const Vector& n = get(normal_map, vt);
+  Eigen::Vector3d vn(n.x(), n.y(), n.z());
+
+  Eigen::Vector3d new_normal = damping_factor * vn + optimized_nvt * vn;
+  new_normal.normalize();
+
+  return Vector{new_normal[0], new_normal[1], new_normal[2]};
+}
+
+enum point_type_t {corner = 3, edge = 1, flat = 2};  // covm
```

Comment:
The commented enum below mentions `// nvt`  which corresponds to the function name.  Is the right one used ?